gather.go

```diff
@@ -146,46 +147,62 @@ func (a *Agent) gatherCandidates() {
 		a.log.Warnf("Failed to run onCandidateHdlr task: %v\n", err)
 		return
 	}
-	a.gatheringState = GatheringStateComplete
 
+	if err := a.run(func(agent *Agent) {
+		a.gatheringState = GatheringStateComplete
+	}); err != nil {
+		a.log.Warnf("Failed to update gatheringState: %v\n", err)
+		return
+	}
 }
 
 func (a *Agent) gatherCandidatesLocal(networkTypes []NetworkType) {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
 	localIPs := localInterfaces(networkTypes)
+	wg.Add(len(localIPs) * len(supportedNetworks))
 	for _, ip := range localIPs {
 		for _, network := range supportedNetworks {
-			conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
-			if err != nil {
-				a.log.Warnf("could not listen %s %s\n", network, ip)
-				continue
-			}
+			go func(network string, ip net.IP) {
+				conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
+				if err != nil {
+					a.log.Warnf("could not listen %s %s\n", network, ip)
+					wg.Done()
```

Comment:
defer wg.Done() (you call it before any return in this function)

Reply:
done.